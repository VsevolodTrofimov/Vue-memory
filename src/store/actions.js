import config from '@/src/utility/gameConfig'
import { preloadImages } from '@/src/utility/preload'

export const loadDeck = ({commit, dispatch}) => {
  const loading = preloadImages(config.cards.options)
  commit('deckLoaderRegister', loading)

  return new Promise((resovle, reject) => {
    loading
      .then(() => {
        commit('deckLoaded')
        resovle()
      })
      .catch(reject)
  })
}

export const startGame = ({commit, dispatch, getters}) => {
  if(getters.deck.loaded) {
    commit('fillBoard')
    commit('setScore', 0)
    commit('setMatched', 0)
    commit('showAllCards')
    commit('setStage', 'game')
  
    dispatch('commitDelayed', {
      mutation: 'maskAllCards',
      delay: config.initialVisibleTime,
      id: 'startGame-1'
    })

    //for future
    preloadImages(['/static/cards/back.png', '/static/banners/EndGame@2x.png'])
  } else {
    commit('setStage', 'loading')
    return new Promise((resolve, reject) => {
      getters.deck.loading
        .then(() => {
          dispatch('startGame')
          resolve()
        })
        .catch(reject)
    })
  }
}


export const restartGame = ({commit, dispatch}) => {
  dispatch('startGame')
}


export const endGame = ({commit}) => {
  commit('setStage', 'end')
}


export const flipCard = ({commit, dispatch, getters}, card) => {
  commit('applyHiding')
  if((card.flipped && getters.flipped.length !== 2) || card.hidden) return

  if(getters.flipped.length === 1) {
    commit('flipCard', card)
    dispatch('matchCards')
  } else {
    commit('maskAllCards')
    commit('flipCard', card)
  }
}


export const matchCards = ({commit, dispatch, getters}) => {
  if(getters.flipped[0].id === getters.flipped[1].id) {
    dispatch('applyCardMatch')
  } else {
    const diff = config.calcScore.failure(getters.matched, getters.unmatched)
    commit('setScore', getters.score + diff)
  }
}


export const applyCardMatch = ({commit, dispatch, getters}) => {
  const diff = config.calcScore.success(getters.matched, getters.unmatched)
  commit('setScore', getters.score + diff)
  
  commit('setMatched', getters.matched + 2)

  getters.flipped.forEach(card => commit('queueCardToHide', card))

  if(getters.unmatched === 0) {
    dispatch('endGame')
  } else {
    dispatch('commitDelayed', {
      mutation: 'applyHiding',
      delay: config.matchedVisibleTime,
      id: 'applyCardMatch-1'
    })
  }
}


export const commitDelayed = ({commit}, {mutation, delay, id, payload}) => {
  if(typeof id === 'undefined') {
    console.warn(
      `Not passing id parameter to commitDelayed puts you at risk
       of unknowingly resetting needed timeout`
    )
  }


  commit('delayMutation', {
    name: `${mutation}---${id}`,
    cb: () => commit(mutation, payload), // mutations cannot commit
    delay
  })
}