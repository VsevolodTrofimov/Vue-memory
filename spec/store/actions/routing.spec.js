import {snapAction, MockPromise, resetTimetable} from 'vuex-snapshot'

import * as actions from '@/src/store/actions.js'

describe('Actions', () => {
  describe('Store routing actions', () => {
    beforeEach(resetTimetable)
  
    describe('Start game', () => {
      it('Matches "deck loded" snapshot', () => {
        const getters = {
          deck: {loaded: true}
        }
        expect(snapAction(actions.startGame, {getters})).toMatchSnapshot()
      })
  
      it('Matches "waiting for deck to load" snapshot', async done => {
        const getters = {
          deck: {
            loaded: false,
            loading: new MockPromise('deck load')
          }
        }
  
        const run = await snapAction(actions.startGame, {getters}, ['deck load'])
        expect(run).toMatchSnapshot()
        done()
      })
    })
  
    describe('Restart game', () => {
      it('Matches snapshot', () => {
        expect(snapAction(actions.restartGame)).toMatchSnapshot()
      })
    })
  
    describe('End game', () => {
      it('Matches snapshot', () => {
        expect(snapAction(actions.endGame)).toMatchSnapshot()
      })
    })
  })
})
