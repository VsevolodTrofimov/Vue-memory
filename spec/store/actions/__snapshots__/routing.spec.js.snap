// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store routing actions End game Matches snapshot 1`] = `
Array [
  Object {
    "message": "COMMIT: setStage",
    "payload": "end",
  },
]
`;

exports[`Store routing actions Restart game Matches snapshot 1`] = `
Array [
  Object {
    "message": "DISPATCH: startGame",
  },
]
`;

exports[`Store routing actions Start game Matches "deck loded" snapshot 1`] = `
Array [
  Object {
    "message": "COMMIT: fillBoard",
  },
  Object {
    "message": "COMMIT: setScore",
    "payload": 0,
  },
  Object {
    "message": "COMMIT: setMatched",
    "payload": 0,
  },
  Object {
    "message": "COMMIT: showAllCards",
  },
  Object {
    "message": "COMMIT: setStage",
    "payload": "game",
  },
  Object {
    "message": "DISPATCH: commitDelayed",
    "payload": Object {
      "delay": 5000,
      "id": "startGame-1",
      "mutation": "maskAllCards",
    },
  },
]
`;

exports[`Store routing actions Start game Matches "waiting for deck to load" snapshot 1`] = `
Array [
  Object {
    "message": "COMMIT: setStage",
    "payload": "loading",
  },
  Object {
    "message": "RESOLUTION: deck load -> resolve",
  },
  Object {
    "message": "DISPATCH: startGame",
  },
  Object {
    "message": "ACTION RESOLVED",
  },
]
`;
